Fix not Recognizing Disabled Object Mapper

TommyWind opened the pull request:
* Fixes not finding disabled object mapper when using dotted field name
notation
* Closes #39456

TommyWind committed:
Fix not Recognizing Disabled Object Mapper

* Fixes not finding disabled object mapper when using dotted field name
notation
* Closes #39456

elasticmachine commented:
Pinging @elastic/es-search

jtibshirani commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java:
```diff
@@ -79,6 +79,22 @@ public void testFieldDisabled() throws Exception {
+            .startObject("foo").field("enabled", false).endObject()
+            .endObject().endObject().endObject());
+        DocumentMapper mapper = mapperParser.parse("type", new CompressedXContent(mapping));
+        BytesReference bytes = BytesReference.bytes(jsonBuilder()
```
It would be great to test some other types as fields as well, such as array fields, since they go through a different parsing code path than simple values.

jtibshirani reviewed (commented):
Hi @TommyWind, thank you for the PR! It’s great that it is a very focused change and that it's unit tested.

While I think this fix does work with the way parsing is currently structured, to me it is non-obvious and maybe fragile. The method contract for `getMapper` is to return the mapper for the given field, or null if no such mapper is defined. But after this change, the method might sometimes return a parent object mapper that lies on the field path. We then rely on `parseObjectOrField` always being called shortly after `getMapper`, so we can check if the returned object mapper is enabled.

I think it would be clearer to omit the change to `getMapper`, and instead create a separate method that iterates through each path segment and returns whether any object mapper is disabled. If there is a disabled mapper, then we can immediately skip further parsing for that field.

TommyWind committed:
Merge remote-tracking branch 'elastic/master' into 39456

TommyWind committed:
change approach and add test

TommyWind commented:
Thanks @jtibshirani for the review. :)

I added a few more test cases and changed the approach to check for disabled mappers higher up the stack. Take a look when you have some time. :) I hope this is what you had in mind.

TommyWind commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -561,35 +582,33 @@ private static void parseArray(ParseContext context, ObjectMapper parentMapper,
+                parseObject(context, mapper, lastFieldName, paths);
             } else if (token == XContentParser.Token.START_ARRAY) {
-                parseArray(context, mapper, lastFieldName);
-            } else if (token == XContentParser.Token.FIELD_NAME) {
```
This statement is impossible to reach, there shouldn't be a field name as a direct child of an array ever? Removed it so the logic for keeping the paths array is a little simpler to follow.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -82,6 +82,20 @@ ParsedDocument parseDocument(SourceToParse source, MetadataFieldMapper[] metadat
         return parsedDocument(source, context, createDynamicUpdate(mapping, docMapper, context.getDynamicMappers()));
     }
 
+    private static boolean isDisabled(ObjectMapper objectMapper, String[] subfields) {
```
Small comment, but I think `containsDisabledObjectMapper` would be clearer.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -386,24 +400,31 @@ static void parseObjectOrNested(ParseContext context, ObjectMapper mapper) throw
                     throw new MapperParsingException("Field [" + currentFieldName + "] is a metadata field and cannot be added inside"
                         + " a document. Use the index API request parameters.");
                 }
+            } else if (isDisabled(mapper, paths)) {
```
This existing method is structured in a fairly confusing way, but given the way it's structured, I think it'd be clearer to check this as part of the 'field name' block: 

```
if (token == XContentParser.Token.FIELD_NAME) {
    currentFieldName = parser.currentName();
    paths = splitAndValidatePath(currentFieldName);
    if (MapperService.isMetadataField(context.path().pathAsText(currentFieldName))) {
        throw new MapperParsingException("Field [" + currentFieldName + "] is a metadata field and cannot be added inside"
            + " a document. Use the index API request parameters.");
        } else if (isDisabled(mapper, paths)) {
            parser.nextToken();
            parser.skipChildren();
        }
    }
} else if ...
```

jtibshirani reviewed (commented):
I like the new approach, and thanks for adding those additional tests. I'll kick off a full CI build now.

jtibshirani commented:
@elasticmachine test this please

TommyWind committed:
rename method and cleanup

TommyWind commented:
@jtibshirani thanks for another round :) I pushed the changes you suggested.

jtibshirani commented:
@elasticmachine test this please

jtibshirani reviewed (commented):
This looks good to me. Before merging I'm hoping someone else from @elastic/es-search could take a quick look as well, in case I am missing anything.

jpountz reviewed (approved):
The change looks good to me as well. Good catch @TommyWind !

jtibshirani commented:
Great, I will merge and backport to the right branches.

TommyWind commented:
Thanks @jtibshirani and @jpountz :)
